Let version parsers fail on rest len

darinkes opened the pull request:
This way the Versioning::parse method actually jumps to the next parser in the list if the previous one was unable to parse the input.

fosskers commented on src/mess.rs:
```diff
@@ -119,6 +119,10 @@ impl Mess {
             next: next.map(|(s, m)| (s, Box::new(m))),
         };
 
+        if i.len() > 0 {
+            return fail().parse(i);
+        }
```
Almost certainly a better way to do this. I will revisit this on the weekend.

darinkes committed:
feat: parse NodeJS-style messes and require full consumption in Versioning

Support `++` and `+~` as `Mess` separators so NodeJS/Debian versions such
as `6.0.1+~3.0.4+~2.0.0+~1.0.0+~2.0.1-1` and `0.7.0++dfsg2+really.0.6.1-15`
parse as a `Mess` and round-trip losslessly through `Display`.

Require each `Versioning::parse` candidate to consume the entire input via
`all_consuming` instead of teaching every individual parser to fail on a
non-empty remainder. This keeps the public `SemVer`/`Version`/`Mess`
parsers composable (they may still leave a remainder for a downstream
`nom` parser, as documented) while ensuring `Versioning::parse` falls
through to the next candidate when a prefix-only match remains.

As a result a version with more parts than a `SemVer` (e.g. the four-part
`10.1.22000.832`) no longer matches the SemVer prefix and strands the
trailing chunk, which previously made it unusable inside a `Requirement`.

Fixes #33.
